UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs

```diff
@@ -19,468 +12,138 @@ namespace MLAgents
-
-                if (graph[BatchSizePlaceholderName] != null)
-                {
-                    hasBatchSize = true;
-                }
-
-                if ((graph[RecurrentInPlaceholderName] != null) &&
-                    (graph[RecurrentOutPlaceholderName] != null))
-                {
-                    hasRecurrent = true;
-                    var runner = session.GetRunner();
-                    runner.Fetch(graph["memory_size"][0]);
-                    var networkOutput = runner.Run()[0].GetValue();
-                    memorySize = (int) networkOutput;
-                }
-
-                if (graph[VectorObservationPlacholderName] != null)
-                {
-                    hasState = true;
-                }
-
-                if (graph[PreviousActionPlaceholderName] != null)
-                {
-                    hasPrevAction = true;
-                }
-                if (graph["value_estimate"] != null)
+                var config = new InferenceEngineConfig
                 {
-                    hasValueEstimate = true;
-                }
-                if (graph[ActionMaskPlaceholderName] != null)
-                {
-                    hasMaskedActions = true;
-                }
+                    Device = inferenceDevice
+                };
+                _engine = InferenceAPI.LoadModel(model, config);
             }
-
-
-            observationMatrixList = new List<float[,,,]>();
-            texturesHolder = new List<Texture2D>();
-#endif
+            else
+            {
+                _engine = null;
+            }
+            _metaDataLoader = new MetaDataLoader(_engine, brainParameters);
+            _inferenceInputs = _metaDataLoader.GetInputTensors();
+            _inferenceOutputs = _metaDataLoader.GetOutputTensors();
+            _tensorGeneratorInvoker = new TensorGeneratorInvoker(brainParameters, seed);
+            _outputTensorApplierInvoker = new TensorApplierInvoker(brainParameters, seed);
+        }
+        
+        /// <summary>
+        /// Return a list of failed checks corresponding to the failed compatibility checks
+        /// between the Model and the BrainParameters. Note : This does not reload the model.
+        /// If changes have been made to the BrainParameters or the Model, the model must be
+        /// reloaded using GiveModel before trying to get the compatibility checks.
+        /// </summary>
+        /// <returns> The list of the failed compatibility checks between the Model and the
+        /// Brain Parameters</returns>
+        public IEnumerable<string> GetModelFailedChecks()
+        {
+            return (_metaDataLoader != null) ? _metaDataLoader.GetChecks() : new List<string>();
         }
 
-
-        /// Uses the stored information to run the tensorflow graph and generate 
-        /// the actions.
+        /// <inheritdoc />
         protected override void DecideAction()
         {
-#if ENABLE_TENSORFLOW
             base.DecideAction();
```

Comment:
If this is always required, it's an anti-pattern: https://en.wikipedia.org/wiki/Call_super